### PR TITLE
Extcommand

### PR DESCRIPTION
**Pull request by CryTime**
# Pull Request Template

## Description

Please include a short resume of the changes and what is the purpose of PR. Any relevant information should be added to help:
* **QA Team** (Quality Assurance) with tests.
* **reviewers** to understand what are the stakes of the pull request.

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 18.10.x
- [ ] 19.04.x
- [ ] 19.10.x
- [ ] 20.04.x (master)

<h2> How this pull request can be tested ? </h2>

Please describe the **procedure** to verify that the goal of the PR is matched. Provide clear instructions so that it can be **correctly tested**.

Any **relevant details** of the configuration to perform the test should be added.

#### Community contributors & Centreon team

- [ ] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

#### Centreon team only

- [ ] I have made sure that the **unit tests** related to the story are successful.
- [ ] I have made sure that **unit tests cover 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: include/com/centreon/engine/engine_impl.hh [hunk 1]
```diff
@@ -21,6 +21,7 @@ class engine_impl final : public Engine::Service {
   grpc::Status NewThresholdsFile(grpc::ServerContext* context,
                                  const ThresholdsFile* request,
                                  CommandSuccess* response) override;
+  grpc::Status GetNbrHost(grpc::ServerContext* context, const GenericString*, GenericValue*) override;
```
Comment (SylvestreG) [inline include/com/centreon/engine/engine_impl.hh:24]: Your rpc should take an empty proto.

File: src/cce_rpc/engine.proto [hunk 2]
```diff
@@ -9,13 +10,21 @@ package com.centreon.engine;
 service Engine {
   rpc GetVersion(google.protobuf.Empty) returns (Version) {}
   rpc GetStats(google.protobuf.Empty) returns (Stats) {}
-
+  rpc GetNbrHost(GenericString) returns (GenericValue) {}
```
Comment (SylvestreG) [inline src/cce_rpc/engine.proto:13]: same here

File: src/centengine/engine_impl.cc [hunk 3]
```diff
@@ -93,6 +97,32 @@ grpc::Status engine_impl::ProcessServiceCheckResult(
   return grpc::Status::OK;
 }
 
+grpc::Status engine_impl::GetNbrHost(grpc::ServerContext* context, const GenericString* request, GenericValue* response) {
+	std::promise<int> p;
+	std::future<int> f1 = p.get_future();
+	//int val = -1;
```
Comment (SylvestreG) [inline src/centengine/engine_impl.cc:103]: you can remove the comment it is not used.

File: src/centengine/engine_impl.cc [hunk 4]
```diff
@@ -93,6 +97,32 @@ grpc::Status engine_impl::ProcessServiceCheckResult(
...
+
+	std::cout << "from server" << std::endl;
+	
+	auto lambda = [&p]() -> int { 
+		p.set_value(host::hosts.size()); 
+		std::cout << host::hosts.size() << std::endl; 
+		return 0;
+	}; 
+
+	//auto fn = std::bind(lambda);
```
Comment (SylvestreG) [inline src/centengine/engine_impl.cc:114]: same here.

File: src/centengine/engine_impl.cc [hunk 5]
```diff
@@ -93,6 +97,32 @@ grpc::Status engine_impl::ProcessServiceCheckResult(
...
+
+	command_manager::instance().enqueue(lambda);
+	
+	//while (val == -1)  {}
```
Comment (SylvestreG) [inline src/centengine/engine_impl.cc:118]: same here

File: test/engine/enginerpc/enginerpc.cc [hunk 6]
```diff
@@ -158,6 +161,32 @@ TEST_F(EngineRpc, ProcessServiceCheckResult) {
...
+
+  enginerpc erpc("0.0.0.0", 40001);
+  auto output = execute("GetNbrHost");
+ 
+  continuerunning = false;
+  th.join();
+
+  std::cout << "test" << std::endl;
+  std::cout << output.back() << std::endl;
+  std::cout << "test" << std::endl;
```
Comment (SylvestreG) [inline test/engine/enginerpc/enginerpc.cc:185]: these cout are not needed anymore

File: test/engine/enginerpc/enginerpc.cc [hunk 7]
```diff
@@ -158,6 +161,32 @@ TEST_F(EngineRpc, ProcessServiceCheckResult) {
   erpc.shutdown();
 }
 
+TEST_F(EngineRpc, GetNbrHost) {
+  std::ostringstream oss;
```
Comment (SylvestreG) [inline test/engine/enginerpc/enginerpc.cc:165]: for thhe moment you test test nothing. Check your rpc return ! :)

File: src/cce_rpc/engine.proto [hunk 8]
```diff
@@ -9,13 +10,27 @@ package com.centreon.engine;
...
   rpc GetStats(google.protobuf.Empty) returns (Stats) {}
-
+  rpc GetNbrHost(google.protobuf.Empty) returns (GenericValue) {}
+  rpc GetNbrContact(google.protobuf.Empty) returns (GenericValue) {}
+  rpc GetNbrService(google.protobuf.Empty) returns (GenericValue) {}
+  rpc GetNbrServiceGroup(google.protobuf.Empty) returns (GenericValue) {}
+  rpc GetNbrContactGroup(google.protobuf.Empty) returns (GenericValue) {}
+  rpc GetNbrHostGroup(google.protobuf.Empty) returns (GenericValue) {}
+  rpc GetNbrServiceDependencies(google.protobuf.Empty) returns (GenericValue) {}
+  rpc GetNbrHostDependencies(google.protobuf.Empty) returns (GenericValue) {}
```
Comment (bouda1) [inline src/cce_rpc/engine.proto:20]: I'm sorry :laughing: 
I don't like those names. They look very frenchy.
GetHostsCount() looks better

File: src/centengine/engine_impl.cc [hunk 9]
```diff
@@ -66,6 +76,143 @@ grpc::Status engine_impl::GetStats(grpc::ServerContext* /*context*/,
...
+  std::promise<int> p;
+  std::future<int> f1 = p.get_future();
+  
+  auto lambda = [&p]() -> int {
+    p.set_value(contact::contacts.size()); 
+	return 0;
+  }; 
+
+  command_manager::instance().enqueue(lambda);
+
```
Comment (bouda1) [inline src/centengine/engine_impl.cc:106]: You can keep this as is but, why not this
command_manager::instance().enqueue([&p]() -> int {
  p.set_value...
  return 0;
});

File: test/engine/enginerpc/client.cc [hunk 10]
```diff
@@ -134,5 +246,54 @@ int main(int argc, char** argv) {
     status = client.NewThresholdsFile(tf) ? 0 : 5;
     std::cout << "NewThresholdsFile: " << status << std::endl;
   }
+  else if (strcmp(argv[1], "GetNbrHost") == 0) {
```
Comment (bouda1) [inline test/engine/enginerpc/client.cc:249]: It is better to use strncmp, more secure
strncmp(argv[1], "GetNbrHost", strlen("GetNbrHost"));

File: test/engine/enginerpc/enginerpc.cc [hunk 11]
```diff
@@ -150,6 +153,182 @@ TEST_F(EngineRpc, GetVersion) {
...
+	while(continuerunning) {
+	  command_manager::instance().execute();
+	  std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    }
+  };
+
+  std::thread th(fn);
+  auto output = execute("GetNbrHost");
+  continuerunning = false;
+  th.join();
```
Comment (bouda1) [inline test/engine/enginerpc/enginerpc.cc:172]: Don't you want to see the thread running while it has not its result?

I have the feeling that the main thread will stop the thread when needed...
Synchronization with tempo is not a good thing

File: test/engine/enginerpc/enginerpc.cc [hunk 12]
```diff
@@ -150,6 +153,182 @@ TEST_F(EngineRpc, GetVersion) {
   erpc.shutdown();
 }
 
+TEST_F(EngineRpc, GetNbrHost) {
+  std::ostringstream oss;
+  oss << "1";
```
Comment (bouda1) [inline test/engine/enginerpc/enginerpc.cc:158]: Very strange

File: script/ci/ut.sh [hunk 13]
```diff
@@ -62,3 +69,4 @@ fi
 exit 0
 
 
+>>>>>>> 73c123c00a38c6b48f219f9389d448d2cb8e05ca
```
Comment (bouda1) [inline script/ci/ut.sh:72]: A bad merge here!

File: test/engine/enginerpc/enginerpc.cc [hunk 14]
```diff
@@ -150,6 +153,182 @@ TEST_F(EngineRpc, GetVersion) {
   erpc.shutdown();
 }
 
+TEST_F(EngineRpc, GetHostsCount) {
+ std::ostringstream oss;
+  oss << "1";
```
Comment (bouda1) [inline test/engine/enginerpc/enginerpc.cc:158]: stringstream should be avoid when not necessary. Here, I think it is not very useful.
You can remove those two lines.

File: test/engine/enginerpc/enginerpc.cc [hunk 15]
```diff
@@ -150,6 +153,182 @@ TEST_F(EngineRpc, GetVersion) {
...
+      std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    }
+  };
+
+  std::thread th(fn);
+  auto output = execute("GetHostsCount");
+  continuerunning = false;
+  th.join();
+
+  ASSERT_EQ(output.back(), oss.str());
```
Comment (bouda1) [inline test/engine/enginerpc/enginerpc.cc:174]: instead of oss.str(), just "1" should do the work

File: test/engine/enginerpc/enginerpc.cc [hunk 16]
```diff
@@ -150,6 +153,182 @@ TEST_F(EngineRpc, GetVersion) {
...
+  erpc.shutdown();
+}
+
+TEST_F(EngineRpc, GetContactsCount) {
+  std::ostringstream oss;
+  oss << "1";
+  enginerpc erpc("0.0.0.0", 40001);
+  std::atomic<bool> continuerunning(true);
```
Comment (bouda1) [inline test/engine/enginerpc/enginerpc.cc:182]: Same remark here and for the following tests.

**Pull request merged**